src/browser.cc

```diff
@@ -0,0 +1,182 @@
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*******************************************************************************/
+
+#include "helpers.h"
+
+#include "insights.h"
+CEREAL_REGISTER_TYPE(insight::MutualInformation);
+
+#include "html.h"
+
+#include "../../Current/Bricks/dflags/dflags.h"
+#include "../../Current/Bricks/file/file.h"
+#include "../../Current/Bricks/net/api/api.h"
+#include "../../Current/Bricks/strings/util.h"
+#include "../../Current/Bricks/waitable_atomic/waitable_atomic.h"
+
+DEFINE_int32(port, 3000, "Port to spawn the browser on.");
+DEFINE_string(route, "/", "The route to serve the browser on.");
+DEFINE_string(output_uri_prefix, "http://localhost", "The prefix for the URI-s output by the server.");
+
+DEFINE_string(input, "data/insights.json", "Path to the file containing the insights to browse.");
+
+using bricks::strings::Printf;
+using bricks::strings::FromString;
+using bricks::strings::ToString;
+using bricks::FileSystem;
+using bricks::WaitableAtomic;
+
+struct TopLevelResponse {
+  std::string SMART_BROWSE_URI;
+  size_t total;
+  std::string HTML_BROWSE_URI;
+  std::string browse_uri;
+  std::string browse_all_uri;
+  TopLevelResponse(size_t i) : total(i) {
+    if (total) {
+      SMART_BROWSE_URI = FLAGS_output_uri_prefix + "/smart";
+      HTML_BROWSE_URI = FLAGS_output_uri_prefix + "/?id=1&html=yes";
+      browse_uri = FLAGS_output_uri_prefix + "/?id=1";
+      browse_all_uri = FLAGS_output_uri_prefix + "/?id=all";
+    }
+  }
+  template <typename A>
+  void serialize(A& ar) {
+    ar(CEREAL_NVP(SMART_BROWSE_URI),
+       CEREAL_NVP(HTML_BROWSE_URI),
+       CEREAL_NVP(total),
+       CEREAL_NVP(browse_uri),
+       CEREAL_NVP(browse_all_uri));
+  }
+};
+
+struct InsightResponse {
+  std::string prev_uri;
+  std::string next_uri;
+  std::set<std::string> tags;
+  double score;
+  std::string description;
+  std::unique_ptr<insight::AbstractBase> insight;
+  InsightResponse(const InsightsOutput& input, size_t index) {
+    if (index) {
+      prev_uri = FLAGS_output_uri_prefix + "/?id=" + ToString(index);  // It's 1-based in the URI.
```

Comment:
`ToString(index);` should be `(index + 1)` if `index` is zero-based (from the condition above I assume it is).